core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -892,3 +871,219 @@ object ZkPath {
+  private class CreateCallback extends StringCallback {
+    def processResult(rc : Int,
+                      path : String,
+                      ctx : Object,
+                      name : String) {
+      Code.get(rc) match {
+        case Code.OK => {
+          // check that exists and wait
+           checkAndWatch
+           setResult(Code.OK)
+        }
+        case Code.CONNECTIONLOSS => {
+          // try again
+          createAndWatch
+        }
+        case Code.NONODE => {
+          error("No node for path %s (could be the parent missing)".format(path))
+          setResult(Code.NONODE)
+        }
+        case Code.NODEEXISTS => {
+          zkHandle.getData(path, false, getDataCallback, null)
+        }
+        case Code.SESSIONEXPIRED => {
+          error("Session has expired while creating %s".format(path))
+        }
+        case _ => {
+          info("ZooKeeper event while creating registration node %s %s".format(path, Code.get(rc)))
+          setResult(Code.get(rc))
+        }
+      }
+    }
+  }
+
+  private class GetDataCallback extends DataCallback {
+      def processResult(rc : Int,
+                 path : String,
+                 ctx : Object,
+                 readData: Array[Byte],
+                 stat : Stat) {
+        Code.get(rc) match {
+          case Code.OK => {
+                  if (checker(ZKStringSerializer.deserialize(readData).asInstanceOf[String], expectedCallerData)) {
+                    info("An ephemeral node [%s] at %s already exists ".format(data, path)
+                      + "it might be because of a past session I have had or connection issues")
+                    checkAndWatch
+                    setResult(Code.OK)
+                } else {
+                    info("An ephemeral node [%s] at %s already exists ".format(data, path)
+                      + "and has been created by someone else")
+                    setResult(Code.NODEEXISTS)
+                }
+          }
+          case Code.NONODE => {
+            info("The ephemeral node [%s] at %s has gone away while reading it, ".format(data, path))
+            setResult(Code.NONODE)
+          }
+          case _ => {
+            setResult(Code.get(rc))
+          }
+        }
+      }
+  }
+
+  private class EphemeralWatcher extends Watcher {
+    def process(event : WatchedEvent) {
+      // if node deleted, then recreate it
+      if(!stop && event.getType == Watcher.Event.EventType.NodeDeleted)
+        createAndWatch
+    }
+  }
+
+  private class ExistsCallback extends StatCallback {
+    def processResult(rc : Int,
+                 path : String,
+                 ctx : Object,
+                 stat : Stat) {
+      Code.get(rc) match {
+        case Code.OK => {}
+        case Code.CONNECTIONLOSS => {
+          // Backoff and try again
+          checkAndWatch
+        }
+        case Code.SESSIONEXPIRED => {
+          error("Session has expired while creating %s".format(path))
```

Comment:
In general, a session expiration should be treated as a condition that prevents the server from making progress, that's the main reason for making it an error. I can agree to make it a warn because it there are recovery mechanisms to prevent to service as whole from coming to a halt or even becoming inconsistent. I don't think it should be info, though. It is important that you get notified that sessions are expiring.